### PR TITLE
.github: Add initial linux CI workflow

### PR DESCRIPTION
**Pull request by samestep**
This is a commandeer of #54091.

TODO:

- [x] understand why the build is [failing](https://github.com/pytorch/pytorch/pull/55176/checks?check_run_id=2254742265) here when it was [succeeding](https://github.com/pytorch/pytorch/pull/54091/checks?check_run_id=2177844748) on #54091
- [x] fix the build failure
- [x] fix the test failure(s)
- [x] add CI check to generate YAML workflows from templates, similar to #55171
- [ ] uncomment the rest of the matrix

### CODE REVIEW

File: .github/templates/linux_ci_workflow.yml.in [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            -e IN_CI \
+            --security-opt seccomp=unconfined \
+            --cap-add=SYS_PTRACE \
+            --tty \
+            --user jenkins \
+            -v ${GITHUB_WORKSPACE}:/var/lib/jenkins/workspace \
+            -w /var/lib/jenkins/workspace \
+            ${DOCKER_IMAGE} \
+            sh -c 'sudo chown -R jenkins . && pip install dist/*.whl && .jenkins/pytorch/test.sh'
+      - name: Report test results
```
Comment (seemethere) [inline .github/templates/linux_ci_workflow.yml.in:132]: You might want to remove this step, I had added it in but it doesn’t actually work correctly from what I can tell

Comment (samestep) [reply .github/templates/linux_ci_workflow.yml.in:132]: Ah yep, I had messaged you on Friday to ask whether to remove this :) will do!

**Comment by seemethere** [comment]
So looks like we're now able to execute tests with GPU and CPU but we're running into issues:

<details>
<summary> GPU </summary>

https://github.com/pytorch/pytorch/runs/2336334319?check_suite_focus=true

```
Tue, 13 Apr 2021 19:38:05 GMT
======================================================================
Tue, 13 Apr 2021 19:38:05 GMT
ERROR [0.034s]: test_data_parallel (__main__.TestDataParallel)
Tue, 13 Apr 2021 19:38:05 GMT
----------------------------------------------------------------------
Tue, 13 Apr 2021 19:38:05 GMT
Traceback (most recent call last):
Tue, 13 Apr 2021 19:38:05 GMT
  File "distributed/test_data_parallel.py", line 350, in test_data_parallel
Tue, 13 Apr 2021 19:38:05 GMT
    out = dp.data_parallel(l, i, dev_id)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 229, in data_parallel
Tue, 13 Apr 2021 19:38:05 GMT
    replicas = replicate(module, used_device_ids)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/replicate.py", line 91, in replicate
Tue, 13 Apr 2021 19:38:05 GMT
    param_copies = _broadcast_coalesced_reshape(params, devices, detach)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/replicate.py", line 71, in _broadcast_coalesced_reshape
Tue, 13 Apr 2021 19:38:05 GMT
    tensor_copies = Broadcast.apply(devices, *tensors)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/_functions.py", line 23, in forward
Tue, 13 Apr 2021 19:38:05 GMT
    outputs = comm.broadcast_coalesced(inputs, ctx.target_gpus)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/comm.py", line 58, in broadcast_coalesced
Tue, 13 Apr 2021 19:38:05 GMT
    return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
Tue, 13 Apr 2021 19:38:05 GMT
RuntimeError: NCCL Error 2: unhandled system error
Tue, 13 Apr 2021 19:38:05 GMT

Tue, 13 Apr 2021 19:38:05 GMT
======================================================================
Tue, 13 Apr 2021 19:38:05 GMT
ERROR [0.022s]: test_data_parallel_model_device (__main__.TestDataParallel)
Tue, 13 Apr 2021 19:38:05 GMT
Test device[0] check at forward time.
Tue, 13 Apr 2021 19:38:05 GMT
----------------------------------------------------------------------
Tue, 13 Apr 2021 19:38:05 GMT
Traceback (most recent call last):
Tue, 13 Apr 2021 19:38:05 GMT
  File "distributed/test_data_parallel.py", line 271, in test_data_parallel_model_device
Tue, 13 Apr 2021 19:38:05 GMT
    test(l.cuda(1), None, inp_cuda1, [1, 0], should_fail=False)
Tue, 13 Apr 2021 19:38:05 GMT
  File "distributed/test_data_parallel.py", line 259, in test
Tue, 13 Apr 2021 19:38:05 GMT
    dpm(inp)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/modules/module.py", line 1015, in _call_impl
Tue, 13 Apr 2021 19:38:05 GMT
    return forward_call(*input, **kwargs)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 166, in forward
Tue, 13 Apr 2021 19:38:05 GMT
    replicas = self.replicate(self.module, self.device_ids[:len(inputs)])
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 171, in replicate
Tue, 13 Apr 2021 19:38:05 GMT
    return replicate(module, device_ids, not torch.is_grad_enabled())
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/replicate.py", line 91, in replicate
Tue, 13 Apr 2021 19:38:05 GMT
    param_copies = _broadcast_coalesced_reshape(params, devices, detach)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/replicate.py", line 71, in _broadcast_coalesced_reshape
Tue, 13 Apr 2021 19:38:05 GMT
    tensor_copies = Broadcast.apply(devices, *tensors)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/_functions.py", line 23, in forward
Tue, 13 Apr 2021 19:38:05 GMT
    outputs = comm.broadcast_coalesced(inputs, ctx.target_gpus)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/comm.py", line 58, in broadcast_coalesced
Tue, 13 Apr 2021 19:38:05 GMT
    return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
Tue, 13 Apr 2021 19:38:05 GMT
RuntimeError: NCCL Error 2: unhandled system error
Tue, 13 Apr 2021 19:38:05 GMT

Tue, 13 Apr 2021 19:38:05 GMT
======================================================================
Tue, 13 Apr 2021 19:38:05 GMT
ERROR [0.013s]: test_data_parallel_sparse (__main__.TestDataParallel)
Tue, 13 Apr 2021 19:38:05 GMT
----------------------------------------------------------------------
Tue, 13 Apr 2021 19:38:05 GMT
Traceback (most recent call last):
Tue, 13 Apr 2021 19:38:05 GMT
  File "distributed/test_data_parallel.py", line 377, in test_data_parallel_sparse
Tue, 13 Apr 2021 19:38:05 GMT
    out = dp.data_parallel(l, i, dev_id)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/data_parallel.py", line 229, in data_parallel
Tue, 13 Apr 2021 19:38:05 GMT
    replicas = replicate(module, used_device_ids)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/replicate.py", line 91, in replicate
Tue, 13 Apr 2021 19:38:05 GMT
    param_copies = _broadcast_coalesced_reshape(params, devices, detach)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/replicate.py", line 71, in _broadcast_coalesced_reshape
Tue, 13 Apr 2021 19:38:05 GMT
    tensor_copies = Broadcast.apply(devices, *tensors)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/_functions.py", line 23, in forward
Tue, 13 Apr 2021 19:38:05 GMT
    outputs = comm.broadcast_coalesced(inputs, ctx.target_gpus)
Tue, 13 Apr 2021 19:38:05 GMT
  File "/opt/conda/lib/python3.6/site-packages/torch/nn/parallel/comm.py", line 58, in broadcast_coalesced
Tue, 13 Apr 2021 19:38:05 GMT
    return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
Tue, 13 Apr 2021 19:38:05 GMT
RuntimeError: NCCL Error 2: unhandled system error
Tue, 13 Apr 2021 19:38:05 GMT

Tue, 13 Apr 2021 19:38:05 GMT
----------------------------------------------------------------------
```

</details>

**Comment by seemethere** [comment]
Looking to resolve disk size issues here: https://github.com/pytorch/pytorch/pull/55955

**Comment by seemethere** [comment]
Successful run: https://github.com/pytorch/pytorch/runs/2355893087

File: .github/scripts/generate_linux_ci_workflows.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            output_file.write('\n')
+        return output_file_path
+
+
+WORKFLOWS = [
+    PyTorchLinuxWorkflow(
+        build_environment="pytorch-linux-xenial-py3.6-gcc5.4",
+        docker_image_base=f"{DOCKER_REGISTRY}/pytorch/pytorch-linux-xenial-py3.6-gcc5.4",
+    ),
+    # PyTorchLinuxWorkflow(
```
Comment (walterddr) [inline .github/scripts/generate_linux_ci_workflows.py:46]: Add a todo for these?

File: .github/scripts/report_git_status.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+CHANGES=$(git status --porcelain)
+echo "$CHANGES"
+git diff
+[ -z "$CHANGES" ]
```
Comment (walterddr) [inline .github/scripts/report_git_status.sh:5]: i am not sure what this does. does it control the return code?

**Review by walterddr — APPROVED** [review]

**Review by malfet — APPROVED** [review]

File: .github/scripts/report_git_status.sh (hunk 3 above)
Comment (samestep) [reply .github/scripts/report_git_status.sh:5]: you mean the last line specifically? it makes the workflow fail if the Git status is nonempty

File: .jenkins/pytorch/macos-test.sh [hunk 4]
```diff
@@ -51,7 +51,11 @@ test_python_all() {
   export GLOO_SOCKET_IFNAME=lo0
   echo "Ninja version: $(ninja --version)"
 
-  if [ -n "$CIRCLE_PULL_REQUEST" ]; then
+  # Try to pull value from CIRCLE_PULL_REQUEST first then GITHUB_HEAD_REF second
+  # CIRCLE_PULL_REQUEST comes from CircleCI
+  # GITHUB_HEAD_REF comes from Github Actions
+  IN_PULL_REQUEST=${CIRCLE_PULL_REQUEST:-${GITHUB_HEAD_REF:-}}
+  if [ -n "$IN_PULL_REQUEST" ]; then
```
Comment (malfet) [inline .jenkins/pytorch/macos-test.sh:58]: We don't really need this change, do we? (As we have not enabled GHA testing on Mac)

File: .jenkins/pytorch/test.sh [hunk 5]
```diff
@@ -274,6 +290,19 @@ test_custom_backend() {
...
+    CUSTOM_OP_TEST="$PWD/test/custom_operator"
+    python --version
+    SITE_PACKAGES="$(python -c 'from distutils.sysconfig import get_python_lib; print(get_python_lib())')"
+    mkdir "$CUSTOM_OP_BUILD"
+    pushd "$CUSTOM_OP_BUILD"
+    cmake "$CUSTOM_OP_TEST" -DCMAKE_PREFIX_PATH="$SITE_PACKAGES/torch" -DPYTHON_EXECUTABLE="$(which python)"
+    make VERBOSE=1
+    popd
+    assert_git_not_dirty
+
```
Comment (malfet) [inline .jenkins/pytorch/test.sh:305]: Would be nice to land those changes as separate PR (i.e. part of ghstack)

File: .jenkins/pytorch/macos-test.sh (hunk 4 above)
Comment (seemethere) [reply .jenkins/pytorch/macos-test.sh:58]: I think having these changes is still good since they really shouldn't affect functionality at all

File: .jenkins/pytorch/test.sh (hunk 5 above)
Comment (seemethere) [reply .jenkins/pytorch/test.sh:305]: I would, but getting this to pass in the first place was a challenge and this shouldn't really affect the tests / builds

File: .github/scripts/report_git_status.sh (hunk 3 above)
Comment (walterddr) [reply .github/scripts/report_git_status.sh:5]: yup. exactly what i meant for return code :-)

File: .github/templates/linux_ci_workflow.yml.in [hunk 6]
```diff
@@ -0,0 +1,174 @@
+# @generated by .github/scripts/generate_linux_ci_workflows.py, Do not update manually
```
Comment (driazati) [inline .github/templates/linux_ci_workflow.yml.in:1]: This file itself isn't generated, right? The script should add in the bit about how it's a generated file to the `workflow/*.yml` instead of putting it here in the source

Comment (seemethere) [reply .github/templates/linux_ci_workflow.yml.in:1]: I'll do this in a follow up PR to this

**Comment by ilia-cher** [comment]
seems like an issue with pytorch_linux_xenial_py3_clang5_asan_build and pytorch_linux_bionic_py3_6_clang9_noarch_test ?
https://ezyang.github.io/pytorch-ci-hud/build/pytorch-master
